### DISCUSSION TITLE
Questions for DAB

### DISCUSSION THREAD

**Discussion by qiaohaojie** [Azure/data-api-builder#1826] [General]
DAB is exciting! 

Some questions:
1. Where and how do I specify custom status_code and exception messages in DAB? For example, I have a stored proc to read data and I want to return status_code 404 if the data not found. And in writing scenario, if my stored proc detect validation error, I want to return status_code 500 (Internal Error; or maybe even more customised user friendly message). 

2. How to I protect the exposed DAB REST end point? How do I provide features like rate limiting, IP restriction etc? Also, as mentioned in question 1, if I can use custom status_code and exception message, how can I pass this to API gateway? 

My initial thought might be sth like:  client --> Azure API management --> DAB --> Azure SQL (stored proc).  Let's say my stored proc decides to return status_code 400 (bad request) with exception message ("input value can not be less than 100"). How my store proc to pass to DAB, and DAB pass to Azure API management, so that at the end, the client will receive this 400 error with the exception message?

**Comment by qiaohaojie** [comment]
I'm surprised that no reply yet. 

As shown below, I have a stored proc that leverages rich Json capability in Azure SQL:

![image](https://github.com/Azure/data-api-builder/assets/3251855/66bdab9f-9035-4aa5-908a-8a6243850476)


My hope is that I can leverage DAB to expose this as a customer-facing REST API. I need the ability to surface custom message ( most cases are complex validation messages when doing PUT/PATCH) .

**Comment by qiaohaojie** [comment]
@yorek , hope this is possible?

**Comment by seantleonard** [comment]
Hi @qiaohaojie , Thank you for reaching and sorry about the delay in response. My answers to your questions:

1. We don't currently support custom status codes or messages for DAB's REST endpoints. This is a useful feature request, and we are now tracking this with #1844
    1. I notice that your stored procedure defines output parameters. DAB doesn't currently support SP output parameters at this time. See #1843 
2. To address rate limiting and other API protections you mention, evaluating an API gateway like you mention is a good route to pursue.

**Reply by qiaohaojie** [reply]
Hi @seantleonard,
Thanks for the reply. It clarifies the answer I need.

Given that Azure SQL now has rich capabilities to handle JSON, feature #1844 would greatly close the gap between advanced data validation and DAB. I'm looking forward to the addition of this feature.

Regarding the output parameter, I think it is important to have it combined with feature #1844. Outputting the dataset in JSON format gives the flexibility to wrap small-but-dynamic datasets (e.g., a single product, given that each product's attributes vary). My client software will be able to render this JSON properly in the UI. All I need is to use DAB to retrieve this JSON text.

Hope my dream comes true soon :)

Regards,
George

**Reply by JerryNixon** [reply]
This recommendation is awesome. I have added it to the product roadmap. A terrific idea.

**Reply by mortenlange** [reply]
Hi Jerry,
Any updates on this?
It would be great to be able to leverage TSQL FOR JSON PATH formatting with DAB.
Right know it seems that these features oppose each other...

**Comment by ksgsharma** [comment]
DAB is awesome! We’ve implemented it with our Azure SQL Server, and our users are loving it!

Any ETA on version 1.8? We’re really looking forward to these new REST features! 😊

- Add regex operator: $filter=col regex pattern
- Add contains operator: $filter=col contains fragment
- Add in operator: $filter=col in (value1, value2)
- Support for new $count aggregate-value keyword
- Support for new $value scalar-value keyword

**Reply by JerryNixon** [reply]
:)

**Comment by JerryNixon** [comment]
This remains on our backlog without any immediate dates for implementation.
